addons/binding/org.openhab.binding.bluetooth.ruuvitag/src/main/java/org/openhab/binding/bluetooth/ruuvitag/handler/RuuviTagHandler.java

```diff
@@ -0,0 +1,163 @@
+                        ruuvitagData.getBatteryVoltage(), SmartHomeUnits.VOLT);
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_DATA_FORMAT,
+                        ruuvitagData.getDataFormat(), null);
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_HUMIDITY,
+                        ruuvitagData.getHumidity(), SmartHomeUnits.PERCENT);
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_MEASUREMENT_SEQUENCE_NUMBER,
+                        ruuvitagData.getMeasurementSequenceNumber(), SmartHomeUnits.ONE);
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_MOVEMENT_COUNTER,
+                        ruuvitagData.getMovementCounter(), SmartHomeUnits.ONE);
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_PRESSURE,
+                        ruuvitagData.getPressure(), SIUnits.PASCAL);
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_TEMPERATURE,
+                        ruuvitagData.getTemperature(), SIUnits.CELSIUS);
+                fieldPresent |= updateStateIfLinked(RuuviTagBindingConstants.CHANNEL_ID_TX_POWER,
+                        ruuvitagData.getTxPower(), DECIBEL_MILLIWATTS);
+                if (fieldPresent) {
+                    updateStatus(ThingStatus.ONLINE);
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            "Received Ruuvi Tag data but no fields could be parsed");
+                }
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Received bluetooth data which could not be parsed to any known Ruuvi Tag data formats");
+            }
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Received bluetooth data with empty ");
+        }
+        super.onScanRecordReceived(scanNotification);
+    }
+
+    /**
+     * Update channel state
+     *
+     * Update is not done when value is null.
+     *
+     * @param channel channel id
+     * @param value   value to update
+     * @param unit    unit associated with value, or null when we are updating DecimalType
+     * @return whether the value was present
+     */
+    private <T extends Quantity<T>> boolean updateStateIfLinked(String channel, @Nullable Number value,
+            @Nullable Unit<T> unit) {
+        if (value == null) {
+            return false;
+        }
+        ChannelUID channelUID = getChannelUID(channel);
+        if (!isLinked(channelUID)) {
+            return true;
+        }
+        if (unit == null) {
+            updateState(channelUID, new DecimalType(value.doubleValue()));
```

Comment:
Hmm -- I am not sure I follow...

```
    public DecimalType(double value) {
        this.value = BigDecimal.valueOf(value);
    }
```

Java docs say:

> Translates a double into a BigDecimal, using the double's canonical string representation provided by the Double.toString(double) method.
> **Note**: This is generally the preferred way to convert a double (or float) into a BigDecimal, as the value returned is equal to that resulting from constructing a BigDecimal from the result of using Double.toString(double).

So for doubles using `BigDecimal.valueOf(x)` is "generally preferred". Thus, it makes sense to use this constructor for the double fields of `RuuviMeasurement`. There's no "double conversions", nor "rounding"?

Then we have the `Integer` fields: `dataFormat`, `txPower`, `movementCounter` and `measurementSequenceNumber` -- they are all 32 bit signed integers (`Integer`), thus all accurately re-presentable by 64 bit double ([ref](https://www.mathworks.com/help/matlab/ref/flintmax.html)). There is conversion to double yes (without loss of accuracy), but no rounding.


So I am not sure what extra benefits extra conversion to `String` brings? We could handle Integers explicitly as well

```java
        if (unit == null) {
            if (value instanceof Double) {
                updateState(channelUID, new DecimalType(value.doubleValue()));
            } else {
                updateState(channelUID, new DecimalType(value.intValue()));
            }
        } else {
            updateState(channelUID, new QuantityType<>(value, unit));
        }
```

But I think that the possible performance gains are not worth the increased messiness of the code

Reply:
Ha. I should have looked into the constructor of `DecimalType` too :smile: So probably the round issues might be less of a problem than I assumed, except there is an extra conversion with the current implementation:
`Number` -> `double` -> `Double.toString` -> `BigDecimal`.
With my suggestion it's:
`Number` -> `toString` -> `BigDecimal`.
My point is to avoid the extra conversion to remove the native double and by calling `toString` directly `Integer` types are also taken care of without these being converted to double inbetween. If `DecimalType` would have accepted `Number` it probably would have done the same as  `QuantityType` does (call `toString`). It's not specifically about performance, just some less conversion steps.

Reply:
Thanks for the explanation...

> [...]  current implementation:
> `Number` -> `double` -> `Double.toString` -> `BigDecimal`.

Ah! Now I see...Indeed the number is converted to string in `BigDecimal.valueOf`...

Hmm, if truly want to avoid conversions, should we avoid `BigDecimal.valueOf` and prefer `new BigDecimal(double)`? This is possible with `DecimalType` which exposes constructor accepting `BigDecimal`, but not with `QuantityType` (forcing conversion to `String` internally).

Process would be
`Double` -> `BigDecimal`.
`Integer` -> `BigDecimal`.

(no loss of accuracy since number is kept as floating point and not converted to decimal 

In implementation it would be something like
```java
    // for QuantityType channels
    private <T extends Quantity<T>> boolean updateStateIfLinked(ChannelUID channelUID, @Nullable Number value,
            Unit<T> unit) {
        if (value == null) {
            return false;
        }
        if (!isLinked(channelUID)) {
            return true;
        }
        updateState(channelUID, new QuantityType<>(value, unit));
        return true;
    }
    // for DecimalType channels
    private <T extends Quantity<T>> boolean updateStateIfLinked(ChannelUID channelUID, @Nullable Double value) {
        if (value == null) {
            return false;
        }
        if (!isLinked(channelUID)) {
            return true;
        }
        updateState(channelUID, new DecimalType(value));
        return true;
    }
    // for DecimalType channels
    private <T extends Quantity<T>> boolean updateStateIfLinked(ChannelUID channelUID, @Nullable Integer value) {
        if (value == null) {
            return false;
        }
        if (!isLinked(channelUID)) {
            return true;
        }
        updateState(channelUID, new DecimalType(value));
        return true;
    }
```


However, in `QuantityType`, the comments make me question what is the best way to convert double to `BigDecimal` -- there `String` is used, as adviced in `BigDecimal` docs as well (`This is generally the preferred way to convert a double (or float) into a BigDecimal`)

```java
        // Avoid scientific notation for double
        BigDecimal bd = new BigDecimal(value.toString());
```
I am not sure that the comment refers to -- perhaps you have better idea?

What do you think?

Reply:
Actually, we only have two cases to deal with
- boxed integers without unit (`DecimalType`). No need for string conversion -> makes sense to use `DecimalType(long)` constructor
- boxed integers and doubles _with_ unit (`QuantityType`). `QuantityType` implementation _forces_ us to convert into `String`. If we take implementation of `QuantityType` fixed, the most simple way construct the instance is `QuantityType(Number, Unit)`.

Diff in the above comment is the actual implementation.

Reply:
Yes makes sense. the one without unit is only integer so that can be handled directly an no need for string conversion. The others can best be handled with `QuanityType(Number, Unit)` as you suggest.

Another suggestion. Maybe write:
```
  if (isLinked(channelUID)) {
    updateState(channelUID, new DecimalType(value));
  }
  return true;
```
Since the linked check is just to avoid calling updateState. And since in all cases it returns true. This change better reflects it's purpose.

One other thing. I looked into `super.onScanRecordReceived(scanNotification);` and that always calls `updateStatus` based on the rssi value. So all these checks here, and status updates are always overridden and thus are basically dead code. So maybe move the call to super to the begin of the method. Then the status updates can be left in, except the ONLINE can be removed because that is already handled by the super.

Reply:
Very good. 

The state updates now cleaned up in f9c8ea7, as discussed above.
Status updates changed in 590a417